content/blog/running-self-hosted-agents-in-azure-devops-using-an-azure-ad-access-token.md

```diff
@@ -0,0 +1,140 @@
+---
+id: blog
+title: Running Self Hosted Agents in Azure Devops using an Azure AD Access Token
+description: Running Self Hosted Agents used to only be possible using Personal Access Tokens, which are a chore to rotate, bound to a single person and often way overprovisioned, making them a security risk. Is there a better way? Let's find out.
+summary: Have you always wanted to get away from Personal Access Tokens in Azure DevOps? In this article, we're going to take a look at how we can use Azure AD Access Tokens to run a Self Hosted Agent Pool in Azure DevOps and walk through the entire process step by step. Then, we're going to throw all that work away and automate it using Terraform so we never have to do that by hand again.
+author: Lee Beenen
+tags:
+  - Azure Devops
+  - Azure AD
+  - Microsoft Entra ID
+  - Active Directory
+  - Self Hosted Agents
+  - Terraform
+slug: running-self-hosted-agents-with-ad-access-token
+date: 2023-08-17
+---
+If you're someone who runs any number of Azure Pipelines a day, at one point you might have considered running a self hosted agent pool.
+
+Maybe you wanted to isolate your workload from the rest of your organization as you are getting sick of pipeline build queues filling up right when you are planning a very important release to production. Maybe you have very specific needs in terms of OS, tooling, whatever.
+
+It doesn't really matter why, you (like me) probably had second thoughts when you saw that the recommended way of getting a Self Hosted Agent to connect with your Azure Devops organization is by using a Personal Access Tokens (PAT).
+
+PAT's might be very easy to create, but in my experience, they are a pain to manage. Rotating a PAT doesn't take long (but cannot trivialy be automatically rotated in a non-interactive setting), but they are bound to a single person, making them unreliable to use in any business-critical applications. Also, if you scope them correctly you end up having a lot of them, making a small amount of manual work become an error-prone and tedious operation, making it very attractive to just have a single PAT that is way overprovisioned for what it was actually meant to do.
+
+Luckily, there's another [way](https://devblogs.microsoft.com/devops/introducing-service-principal-and-managed-identity-support-on-azure-devops/). Since late March, Microsoft has made it possible for organizations that use Azure AD (soon to be renamed to Microsoft Entra ID, for reasons still unclear to me and everyone else) to use App Registrations and their resulting Service Principals to create Azure AD Access Tokens that can be used in place of PAT's.
+
+Sounds great! How though? Well that's what we're here for, so let's get going.
+
+{{<raw>}}<h3 class="display-5">Setting things up</h3> {{</raw>}}
```

Comment:
Just checked some of my own blog posts. I tend to use just a markdown level 2 header, so: 

```markdown
## Some section
```

Which results in (strangely enough) not-quite-as-JUMBOTRON headings 😉